### ISSUE TITLE
Importing friendly into IPython appears to break %pprint functionality

### ISSUE THREAD

**Issue by kevinbowen777** [friendly-traceback/friendly#14]
Hello,
I am running into an issue when working within IPython and need to toggle
the pprint functionality either 'off' or 'on'.

Here is the error message I receive when trying to run the "%pprint" magic
command from within IPython:

```Traceback (most recent call last):                                              
  Code block [2], line 1, in <module>                                           
    get_ipython().run_line_magic('pprint', '')                                  
  File "LOCAL:/IPython/core/interactiveshell.py", line 2351, in run_line_magic  
    result = fn(*args, **kwargs)                                                
  File "LOCAL:/decorator.py", line 232, in fun                                  
    return caller(func, *(extras + args), **kw)                                 
  File "LOCAL:/IPython/core/magic.py", line 187, in <lambda>                    
    call = lambda f, *a, **k: f(*a, **k)                                        
  File "LOCAL:/IPython/core/magics/basic.py", line 308, in pprint               
    ptformatter.pprint = bool(1 - ptformatter.pprint)                           
AttributeError: 'BaseFormatter' object has no attribute 'pprint'
```

-----
IPython config information for importing friendly:

```
## lines of code to run at IPython startup.
  c.InteractiveShellApp.exec_lines = [
      "from friendly.ipython import *",
    ]
```
Should I be placing a set_formatter() call here? I tried set_formatter('rich') or 'jupyter' with no success.

I have also tried commenting the above out of my ipython_config, starting 
IPython, and then manually importing with the same results. Additionally,
I tried to limit my import of friendly from a * import to only using:
     `from friendly.ipython import what, where, why`
again with the same error message.

----

Here is my system information:
```
Python 3.9.7 (default, Aug 31 2021, 13:28:12) 
IPython 7.28.0 -- An enhanced Interactive Python. Type '?' for help.
friendly_traceback 0.4.54; friendly 0.4.18.
```

I have confirmed that this also happens with IPython 7.20 & Python 3.8.10.
My distros are Mint 20.2 & Manjaro 21.1.5

Please let me know if you need any additional information, from me. Or, let
me know if I am doing something wrong, myself, to cause this problem.

Thanks for your work on this module. I appreciate it.

**Comment by aroberge** [comment]
I've never used pprint with IPython.  I will need to investigate.

**Comment by kevinbowen777** [comment]
Thanks for the quick reply!
I am in no hurry; but, I am around for a few hours to test some things for you, if you need. Having improved messages is better than nothing at all(though, I do like the rich text formatting). ;-)

As far as using %pprint in IPython, take something like this, for example:
    `[name for name in dir(__builtins__) if name > 'a']`
pprint, when on, will add a '\n' after each item in a list. I don't really use it for much more than saving screen space for long lists.

**Comment by kevinbowen777** [comment]
Toggling %pprint now seems to work as expected. Detailed error messages from friendly also appear to be working for me. Thanks for the _really_ quick response and (partial) resolution!

Hopefully, a solution for restoring color to the messages via rich is possible. Feel free to ping me if you want me to look into anything further. I'll check back later tonight, or tomorrow.
Thanks again for being amazingly responsive! Hope the rest of your weekend is quieter.

**Comment by aroberge** [comment]
Keeping this issue open until I can figure out if it is possible to use Rich and toggle pprint (or equivalent).

**Comment by kevinbowen777** [comment]
Would it be worthwhile, for me, to test anything for you with the new version? Happy to help out. Haven't seen any response from the rich project  on your bug report.

**Comment by aroberge** [comment]
I put this in the "Discussion" area of the project, as it did not seem to be suitable to call it a bug as such.

**Comment by kevinbowen777** [comment]
Hmm. I'm not sure that 4.20 is working as expected.

I am using `import friendly.ipython_plain` and am unable to toggle pprint on/off and continue to get this message. I am getting the friendly error messages(obviously not in color).

**Comment by aroberge** [comment]
New version uploaded.

If `%pprint` support is not added in Rich over the next couple of weeks, I will see if I can find a way to disable Rich in IPython once it has been enabled. (I think I know how to do it, in principle, but it would be messy.) If I can do this, then I should be able to implement support for an equivalent to `%pprint` in friendly.python, without having to resort to importing a different module.

**Comment by aroberge** [comment]
@kevinbowen777 I'm going to leave this issue open until I add some relevant information in the documentation. Feel free to comment further, **especially** if you find some bugs! ;-)

**Comment by aroberge** [comment]
With the newest version of Rich, `%pprint` works as expected with IPython.  There is no need to reenable Rich (by calling `dark()` or `light()`) after toggling `%pprint` back on.  Since that is the case, there is no need to document anything.

After uploading a new version of friendly (0.4.25) to Pypi, this issue can now be closed.

**Comment by kevinbowen777** [comment]
I've upgraded both friendly and rich, now when I start IPython, I am getting the following:
```
Traceback (most recent call last):                                              
  Code block [1], line 1, in <module>                                           
    from friendly.ipython import *                                              
AttributeError: '_State' object has no attribute 'exception_before_import'
```
I can confirm that I do not need to re-rerun either dark() or light() after running %pprint again. Let me know if you need additional info from me. Thanks.

**Comment by aroberge** [comment]
On Mon, Nov 8, 2021 at 12:05 AM Kevin Bowen ***@***.***>
wrote:

> I've upgraded both friendly and rich, now when I start IPython, I am
> getting the following:
>
> Traceback (most recent call last):
>   Code block [1], line 1, in <module>
>     from friendly.ipython import *
> AttributeError: '_State' object has no attribute 'exception_before_import'
>
> I can confirm that I do not need to re-rerun either dark() or light()
> after running %pprint again. Let me know if you need additional info from
> me. Thanks.
>

When you start friendly, do you get the following:


friendly-traceback: 0.4.59
friendly: 0.4.25

The attribute 'exception_before_import' is a recent addition to
friendly-traceback (which should have been updated when you updated
friendly).

**Comment by kevinbowen777** [comment]
Hrm. No, I was getting `friendly_traceback 0.4.55; friendly 0.4.25.`

I ran another `pip install --upgrade` specifically for `friendly-traceback` and that resolved the issue.

IIRC, upgrading friendly also pulled in the upgrade for friendly-traceback; but, that doesn't seem to be the case, for me, this time. 
I tried using another venv and upgraded friendly, and I confirmed that I needed to upgrade friendly-traceback separately.
